### ISSUE TITLE
Cannot seem to commit write cache

### ISSUE THREAD

**Issue by tybofu** [radare/radare2#2738]
Hello,
I tried to used the write cache but I ran into the following problem: I cannot seem to commit the write cache to the file.
I created a small text file and I tried the following operation:

```
$ r2 -w -e io.cache=true ./test
[0x00000000]> o
- 6 ./test @ 0x0 ; rw size=6  
[0x00000000]> px
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00000000  9090 9020 210a ffff ffff ffff ffff ffff  ... !...........
0x00000010  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00000020  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00000030  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00000040  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00000050  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00000060  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00000070  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00000080  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00000090  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x000000a0  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x000000b0  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x000000c0  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x000000d0  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x000000e0  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x000000f0  ffff ffff ffff ffff ffff ffff ffff ffff  ................
[0x00000000]> wx 202020
[0x00000000]> wc
idx=0 addr=0x00000000 size=3 909090 -> 202020 (not written)
[0x00000000]> wci
Error writing change at 0x00000000
[0x00000000]> wc+ 0
Error writing change at 0x00000000
[0x00000000]> wc+
Invalidate write cache at 0x00000000
[0x00000000]> wc
idx=0 addr=0x00000000 size=3 909090 -> 202020 (not written)
```

Do you have any idea about what could go wrong? Is it my setup?

Thank you a lot in advance for your help!

**Comment by radare** [comment]
yep. io cache seems broken. i think we have some related tests. but it
will be good
to add some more and use TDD from now on as long as the testsuite is
green and the failing
tests are marked as broken.

thanks for reporting

**Comment by CodingFree** [comment]
I think I am having the same problem:

IO.Cache is already enabled:
```
io.0xff = 255
io.aslr = false
io.autofd = true
io.buffer = false
io.buffer.from = 0
io.buffer.to = 0
io.cache = true
io.enforce = 0
io.ff = true
io.pava = false
io.raw = false
io.sectonly = false
io.va = true
io.vio = false
```

And I have a commit pending:
`idx=0 addr=0x004005c4 size=5 48656c6c6f -> 48616c6c6f (not written)`

However, I am not able to commit the changes into the file:
`Error writing change at 0x004005c4`

I am using the following version:

> radare2 1.2.0-git 13571 @ linux-x86-64 git.1.2.0-git-5-gb2cefac
> commit: b2cefac2ef7d901c296ba2ecf0ec5efe3151eb6c build: 2017-01-23

**Comment by radare** [comment]
cc @condret

if we fix this, should be in siol ulness its quick fix.

**Comment by radare** [comment]
can youa add a test case for r2r?

**Comment by radare** [comment]
as a workaround you can use wc * > fun.r2

and then r2 -i fun.r2 -w bintopatch

**Comment by CodingFree** [comment]
May this be part of the problem?

```
[0x004005c4]> wc
idx=0 addr=0x004005c4 size=5 48656c6c6f -> 48616c6c6f (not written)
[0x004005c4]> wc * > fun.r2
[0x004005c4]> q
root@vps326280:~# cat fun.r2
root@vps326280:~#
```

**Comment by radare** [comment]
sorry, wc* not wc *

**Comment by radare** [comment]
@condret fixed in siol?

**Comment by radare** [comment]
moving up because that depends on the siol branch to be merged

**Comment by MaskRay** [comment]
Can be closed?

Seems to be fixed by commit 1c4407cb69832307a340aa10a71f23cd4cc4997a
